### ISSUE TITLE
v4 goals

### ISSUE THREAD

**Issue by mreinstein** [pixijs/pixi.js#2224]
@englercj we had a brief discussion that was off-topic in #2205 about what might be the goals of pixi v4.

I wanted to start a thread on this, and weigh in on some potential ideas. I'll preface it by saying that I understand many/all of these may not be possible in the next release for a variety of reasons. But I think they are the most interesting from my perspective as a pixi.js user, so I want to get them out before I forget, and get the discussion going. :)

I see several issues tagged with  v3.1 and v3.2. Curious, will we jump straight to v4 or are those releases still planned?

* **memory performance** - overall pixi.js is pretty tight and doesn't leak much memory. Still it might be worthwhile to do a code audit and look for areas to squeeze even more; for example we still use `Array.splice` to insert elements, and there may be other performance improvements around array handling. https://gamealchemist.wordpress.com/2013/05/01/lets-get-those-javascript-arrays-to-work-fast/

* **slight architecture tweak** pixi v3 was awesome in that it modularized the code, but the problem is everything is still linked together via these "glue" objects. e.g., https://github.com/pixijs/pixi.js/blob/master/src/filters/index.js So it's modular but it still pulls in everything + the kitchen sink. Would love to see us just require the modules we need in the appropriate locations instead.

* **lighting**. lighting, lighting, lighting. I don't expect pixi to get real time global illumination with light probes and all other kinds of advanced tech, but this is really where pixi (and to be fair all web focused game engines) are lacking. Shadows, some kind of per-pixel lighting, etc would be amazing. I get this is a *metric tonne* of work, and it wouldn't be cheap. So my expectations here are set that it wouldn't happen for a while. But if you're looking for a hot new feature to add, this is it, easily.

* **cutting edge javascript/browser features** This is a bucket of several hot things coming out which *may* be too soon to include in production pixi but have big potential:

  * **shared array buffers** the ability to share data between worker threads without neutering the array buffer(s) opens up performance possibilities https://www.chromestatus.com/features#shared array buffers
  * **webgl shared resources** firefox has supposedly shipped with the ability to share stuff between 2 webgl contexts. Chrome is working on it: https://www.chromestatus.com/features#shared resources
  * **webgl 2** noone has even shipped this yet, so it's probably way too soon, but being able to support those right out of the gate or being ready to do so would be amazing. Mostly texture compression capabilities and new shader features.
  * **asm.js** Ok so this one would be fucking bananas in terms of effort. But getting crazy with it, porting pixi.js to some kind of typed language supported by emscripten means we could generate an asm.js compatible javascript build that is heavily optimized for performance. Firefox and edge have already shipped implementations, and Chrome is working on it. https://www.chromestatus.com/features#asm.js

To re-iterate..not demanding or asking for any of these. My goal is to just share stuff that is coming down the pipeline in terms of next-gen browsers that sounds like it could be great for pixi users.

Please feel free to proceed to tell me these are crazy and unworkable. :)

**Comment by drkibitz** [comment]
@mreinstein I like your list which I will comment here, after I add a couple that I've rattled on about quite a bit before.

- **Completely abstracted interaction api** PIXI should not be interested in subscribing to the platform specific event implementations, and instead just provide an api to trigger a PIXI event model, which should be a complete abstraction of a common model that would support all platforms (e.g. pointer events, or we can just continue calling them pixi interactions). Speaking more abstractly, the api should more or less be I/O driven.
- **Allowed to run in multiple platform friendly** e.g. isomorphic, this is something that would need to happen with your points on workers and/or an asm.js implementation, even still, this is something that can and should be achieved rather easily (if kept top of mind) outside of those goals.
- **Tests** PLEASE!

As far as your points, my thoughts on them are as follows:
- **memory performance** I'm pretty sure everyone would be game for this, but I agree with a point @englercj has brought up a fews times now, and that this endeavour should probably start with implementing a benchmarking suite. Which I propose should be categories of perf specific to JavaScript, and perf specific to rendering, as it pertains to different renderers (canvas, webgl, etc.)
- **slight architecture tweak** I think at least a few people already agree here, myself included. I had forked PIXI a while ago before V3 trying out this approach (still within a single package). It is totally possible, and IMHO the preferred approach for true modularity that would be appreciated by consumers of PIXI that are not simply "trying it out" or using it as a shared library script. This would require more maintenance, and would probably require more maintainers. This would then probably result in talking about things like semantics, versioning, governance and so on... I'm all for this, but understand the hesitance, we just need to jump in at some point and experience some growing pains. Convenience may try to trump this task so we will need to be clear on the benefits, and come up with an approach that still provides the glue, and an approachable shared library for those completely dependent on this api, and those just testing the PIXI waters.
- **lighting.** lighting, lighting, lighting. Have you seen https://github.com/pixijs/pixi-lights? I haven't used it myself, but @englercj may be able to speak more here. Maybe a good place to start?
- **cutting edge javascript/browser features** I'm not on one side of the fence here or another. I would need to see more clearcut examples of the benefits of each of those features you mentioned, except for asm.js, that one is pretty clear to me ;)

**Comment by drkibitz** [comment]
Hows this one? A common approach to engines out there:

- **Major architectural tweak** Separate the PIXI backend from the frontend. The backend are things related rendering implementations (webgl, canvas), and anything that is specific to a platform for those renderers, or provided for introspection and as a convenience. The frontend is the scene graph, common display objects, timers, and interaction abstraction. Outside of this are plugins, these should be plugins specific to the frontend, or the backend. e.g. A pluggable api on both ends ;)

**Comment by JiDW** [comment]
Is a PIXI 3.1 (or 3.2) version not planned anymore ? There's already a lot of features tagged for a PIXI v3.x right here.

https://github.com/pixijs/pixi.js/issues?q=is%3Aopen+label%3Afeature-request+milestone%3Av3.1
https://github.com/pixijs/pixi.js/issues?utf8=%E2%9C%93&q=is%3Aopen+label%3Afeature-request+milestone%3Av3.2

**Comment by mattstyles** [comment]
*Architecture* - I like the idea of a separation between render-specific tasks (which I suspect is most of PIXI) and render-agnostic tasks (the backend). I think this would pave the way to allowing modular includes of PIXI more easily. I'm guessing the end goal here is to `require` what you like (in CommonJS) and leave out what you dont need? i.e. conditionally add interactions, filters etc when required. I'm not necessarily suggesting a monorepo approach but it almost sounds like its heading that way with the modularisation anyway.

*Next branch* - a next branch that includes experimental features would be good.

*Benchmarking* - agreed that this is essential to implement consistent improvements to perf.

**Comment by ivanpopelyshev** [comment]
My opinion:

* asm.js no-no-no-no-no, buck this, it completely contradicts with pixi approach. asm.js for box2d in separate webworker is totally ok though :) 

* MOAR LIBS! I'm working on pixi-spine and pixi-tilemap, and there'll be pixi-gc for those who just like to use extra canvases and forget to free them.

* MOAR SHADERS! If something is slow, move it to shader. Animated tiles? Wirte a shader. Graphics? Write a shader. Animated SVG-like curves? Write a shader. ParticleContainer doesnt have gl.POINTS implementation, that's a shame!
 
* ShaderLib support - three.js approach is one of the best, why not try it?

* WebGL2 - Transform feedback for spine - WANT!

* Completely abstracted interaction api - Current interaction API sucks. I cant even specify bounding box for container so it wont check its children. So much pain with interaction in pixi-spine! And it completely ignores masks!

**Comment by ivanpopelyshev** [comment]
* Scene Editor - yes, we need it. Basic sprites, meshes, spine animations, and stuff. Even if it will be in form of (Playground)[http://www.html5gamedevs.com/topic/18619-pixi-playground/]

**Comment by mehmetminanc** [comment]
*Migration to ES6* - I am surprised no one mentioned yet, but how about migrating to ES6? I have skimmed the project a little and feel like it'll take a breeze to get there, since the project is already highly modular and OO, which is great. 
Sure this means a complete overhaul but end product  could be a spec-complaint project. When browsers finally support ES6, we may not even need to `browserify` or `babelify`. Well, at least, to a lesser extent.

**Comment by englercj** [comment]
Since it seems like others want to talk about it, I'll go ahead and dump some of my thoughts about what I want to do in v4. A lot of this has been keeping notes ever since we wrote v3, 

## v4 Core Ideas

With v4 I feel like we need to set strong "words to live by" that define all our decisions on the development roadmap for this major version. Here are the core values I think should drive the development of pixi's next version.

1. ES6, wherever possible.
 - Use babel to fill the gaps and transpile to ES5.
2. Modularity, modularity, modularity.
 - Anything that doesn't *have* to be in the core, isn't.
 - No one should think twice about putting core pixi into their project due to weight.
 - The project becomes just a collection of plugins, even if our default bundle includes most of them.
 - Another implication of this is I want a plugable renderer. If you don't use a feature you don't pay for it (like stencil buffer isn't activated unless you use Graphics).
3. WebGL First, including WebGL 2 support and compressed textures.
 - Just like ES6 this is the future, lets make pixi.js the three.js of 2D.
 - No more skipping features because it can't be done in canvas effectively.
 - Canvas rendering moves to a plugin, no longer part of the core.
4. Testable and benchable
 - All code needs to be written in a highly testable, highly benchable way.

## Bundle Builder

Since the project will be so modular, it becomes imperative that we build an easy-to-use custom bundle tool. Users would browse to the build site and select checkboxes for which things they want in their build of pixi, and the tool would build that set and give it to them. There can even be presets that have specific sets of plugins selected for you.

Examples of builders:

- http://jqueryui.com/download/
- http://ckeditor.com/builder

This way we can move things (even commonly used things) to plugins, and if we feel it is appropriate they can be default checked (but uncheckable) on the build tool. This solves the problem people have with pixi (particular game engine consumers) where there are many things in core that they don't use.

## Npm namespace

All v4 modules including the core and the official plugins should live under the `@pixi/` namespace. I claimed the `pixi` user on npm so we can publish our modules to that namespace.

For example, here are some module names we will likely use:

- `@pixi/pixi` - main pixi module, the core.
- `@pixi/interaction` - interaction plugin
- `@pixi/filter-pack` - the filters pack

## Other Thoughts

### Interaction Plugin

This will certainly be a separate plugin, and not shipped in core. It will also follow many of the ideas that @drkibitz has been promoting including being abstracted from DOM, plugable, and easily extensible. This plugin will also include accessibility features.

### Shaders and sharing GLSL

I'd like to make it easier, or more obvious how to integrate custom shaders into Pixi objects. This includes objects that are solely drawn with your shader and extra filters. It should be trivial for anyone to change the drawing behavior of any object. Remember, WebGL First. I'm not 100% how to do this, but I'm considering multiple options.

### Scene Editor

I started a project for this using electron, but never got around to finishing it. Maybe I will make it more pixi specific and open source it.

**Comment by mreinstein** [comment]
@englercj  regarding the ES6 wherever possible, I'm wondering what the impact of that might be: is it largely a re-write of the codebase?

The reason why I ask this is if it's going to be a major body of work to convert, it might be worthwhile to convert to some typed language, and using emscripten to produce an asm.js optimized build. c/c++ are supported out of the box, and any language that uses LLVM *should* work too.

I realize that this would be a major endeavor, but if it's on the same scale of upping to ES6, I think this might be more worth it, as at least with asm.js there's a performance improvement in all the "non-drawy" bits within pixi

**Comment by mehmetminanc** [comment]
@mreinstein On ES6's defense, the changes will mostly be one-to-one conversions, e.g. classes instead of `prototype` class functions, `import` instead of `require`. I would go as far to claim most can be automated. These should not result in any additional mental burden for both the users, who upgrade to, say, v4 and the developers. I don't think this is on the same scale of complete rewrite in a native language. 

Just curious, can SDL be compiled against asm.js with WebGL support? It would fill the niche just right.

**Comment by ivanpopelyshev** [comment]
BTW, first version of compressed textures support is almost ready, Im using code from issue #2200

**Comment by mreinstein** [comment]
> Just curious, can SDL be compiled against asm.js with WebGL support? It would fill the niche just right.

I don't know. I'm sure it's doable on some level. If the unreal and unity engines can be successfully run through emscripten to produce tightly optimized js, I don't see why SDL wouldn't be similarly portable.

Another possibility is rather than trying to convert all of pixi, maybe there are hot spots in the rendering engine that could be pulled out and converted to asm.js modules.

**Comment by englercj** [comment]
> regarding the ES6 wherever possible, I'm wondering what the impact of that might be: is it largely a re-write of the codebase? The reason why I ask this is if it's going to be a major body of work to convert, it might be worthwhile to convert to some typed language, and using emscripten to produce an asm.js optimized build. 

Yes, it may be possible to do something like this but I wouldn't be considering it for v4 it would be an entirely different project. There is also the marshalling project when using an asm.js library that makes it difficult for api-heavy applications to work well with it.

> BTW, first version of compressed textures support is almost ready

That's awesome, I wouldn't get your hopes us about anything written for v3 to work in v4 though. I want to revamp the internals a lot which means major changes.

> Another possibility is rather than trying to convert all of pixi, maybe there are hot spots in the rendering engine that could be pulled out and converted to asm.js modules.

I actually had done this, writting asm.js by hand. The performance was actually much worse due to having to marshall data back and forth between the asm and js layers. After working with some friends at Mozilla we determined it really is all or nothing, changing only small functions actually only makes it slower.

**Comment by mreinstein** [comment]
>  I wouldn't be considering it for v4 it would be an entirely different projec

That seems fair to me. Probably better to come up with a new project rather than trying to massively contort pixi into something that it's not. :)

> we determined it really is all or nothing, changing only small functions actually only makes it slower.

Oh that's really interesting. Thanks for the info!

**Comment by GoodBoyDigital** [comment]
Hey All! 

Thanks for sharing your ideas for the v4! All really great ideas. Here's my 2pense..

### Modules.
Great idea. Its time to split this stuff out into separate repos. Fortunately v3 is quite modular already so I do not think that this transition will be to tricky to achieve. I like the idea of core pixi being extremely light.

I would also be keen to keep ensure that users who would like to use pixi as a full library would not need to concern themselves too much with the split nature. This is the build tool comes in and also including a repo that brings all the main bits together. This is pretty much what @englercj has suggested with one minor tweak:

@pixi/pixi - the build that contains a full featured version of pixi incorporating the 'main' modules.
@pixi/core - main pixi module, the core.
@pixi/interaction - interaction plugin
@pixi/filter-pack - the filters pack 

### WebGL
As @englercj suggested, we should make it webGL first. I am also keen to integrate webGL 2 functionality where we can. I have already been playing around with modifying pixi to use VBOs in another branch. This will be the perfect candidate as it is very easy to build a fall back for it.

I would also be keen to make it much easier to for people to go off-piste and work at a lower level for more advanced users. This includes things like custom shaders and the ability to build custom buffers. 

@ivanpopelyshev, your suggestions about shaders libs is great. The modular nature of v4 will mean we can create lots of tidy little modules that can be specialized (such as the pixi-tilemap).

### Graphics
We should split out the graphics class in to smaller chunks. The classes are pretty massive. We should also consider new graphics rendering methods. Currently the graphics module for pixi works really we for static geometry. I would propose we create new graphics modules for things like rendering high quality lines or dynamic shapes. This way people can big the right tool for the right job.

### Scene editer.
Its great to see that some scene editors are already under construction. Tooling is something that will really benefit users. We are currently working on an open source project with a JIBO and Adobe to allow flash the ability to publish to pixi. This means we will be able to leverage amazing UI of Animate CC (formerly Flash CC). This will put a lot of power back into the animators hands which is rad.

### ES6
If there is no speed penalty of performance issues, then I am all for converting the code to es6 for v4. As mentioned above, it would not be that big of a deal to achieve due to its current structure being a good fit for es6. We should definitely start with converting a small part of pixi and performing some benchmarks before jumping in all the way mind.

### asm.js
As @englercj mentioned, asm.js is fast when it lives within itself. As soon as it needs to mix with js it starts to lose its benefits. If that ever changes we can look at this as a possible solution again. For now we should stick with regular js. 

Exciting times for all! 2016, it will be a good year for Pixi :+1: 

Please keep your thoughts coming!

**Comment by ivanpopelyshev** [comment]
Ha-ha-ha, Im slowpoke: #2243 . This thing belongs to v4 too.

**Comment by photonstorm** [comment]
I can't help but wonder if SVG / CSS ought to play a bigger role here (well, a role at all!) - when you look at the beautiful Christmas games made by Google https://santatracker.google.com/#village which are all svg and Polymer I have to think this is quite important for web games actually.

**Comment by ivanpopelyshev** [comment]
@photonstorm well, as a plugin, SVG and CSS renderer will be awesome. Even filters can work there!

**Comment by ivanpopelyshev** [comment]
By the way, [compressed textures](https://github.com/pixijs/pixi-compressed-textures) are ready. DDS, PVR's are working and resolution can be selected.

**Comment by ivanpopelyshev** [comment]
@photonstorm if we solve problem with modularity, will you use pixi-v4 as phaser-v3 renderer? Or did you make some awesome changes that pixi cant support because of its architecture?

**Comment by photonstorm** [comment]
Sorry but no, in it's current form pixi isn't suitable for use in Lazer. The reason is that Pixi isn't really *just* a renderer at all (and hasn't ever been). It's also not actually modular despite the changes made in v3. Please don't take these as slights against it, they're not meant in that way at all, I'll try and explain:

My reasoning that it's not actually a pure renderer (or modular) is based on the fact that pretty much everything is still deeply coupled, despite being in separate modules. The renderer can only render Pixi Sprites for example, Pixi Sprites rely on the Pixi Matrix for transforms in its own specific way. The whole scene graph is entirely based on *deferred* updates, so you never truly get object positions in real-time (you're always a frame behind), which a pure renderer would never actually do - it would just render exactly what it's given, and it wouldn't even expect to be given a Sprite imho.

There's actually quite a large dependency chain in there when you start breaking it down, and you drag this chain with you where-ever you walk, no matter what. I fully understand why it was built like this, but I honestly think it's as much of a 'renderer' as CreateJS is (it is of course infinitely better than CreateJS will ever be :) )

There is absolutely *nothing* wrong with this approach either. For the most part it's exactly what devs want anyway - but it doesn't fit what I need or where I believe web game dev is heading, sorry. I think StackGL is a good example of how things should break down, but the way they use separate npm modules for everything is a royal pain the ass actually (and not a direction Pixi should head), but the *concept* behind the structure is absolutely bang on the money.

**Comment by ivanpopelyshev** [comment]
@photonstorm ok, got it, you just want to re-implement big number of things, from textures to spine support.

Am I right that on the scale from "we are hardcore GPU users and we compile everything with asm.js"  to "We are fans of CSS and SVG" lazer will be closer to latter than pixi?

**Comment by photonstorm** [comment]
Not really a valid comparison as I don't think Pixi is for hardcore GPU devs *at all* :)

**Comment by englercj** [comment]
@photonstorm Is right, and that issue is something I want to try to address in v4. One of the biggest things holding us back from doing that is how focused we have been on making the library "drop in and everything works magically". The more "drop and play" we make the lib, the less useful it becomes as a library dependency.

I want to try to fix this in v4. I'm hoping we can balance by writing pixi as a highly consumable collection of modules, but still have that "drop and play" style coming out of the build generator. That way framework creators can choose what they need (like just core) and not worry about the build generator, and end users can generate a full build like they use now.

I don't know if we can make it work, but I want to try. StackGL is a good example of what I am shooting for. I'm curious though, you mention you like the modularity but not that they are on npm. How do you expect the modules to be split and distributed instead?

**Comment by ivanpopelyshev** [comment]
Im all for modularity. 

Want extra functions? - add library with its own container, sprite elements, renderers, shaders and stuff. 

Want support for atlases, compressed textures, normal maps? - Use library adds middlewares into loader.

I just dont understand why @photonstorm wants to make his engine separate from this architecture. Im sure that anyone here can make a big game from zero lines of code, using nothing, developing its own architecture. I joined this community to help other coders and use their experience in areas I suck, and now i see that one of most big its supporters says "nah". So if I publish new libraries (tiles renderer or portal engine), I'll need to write plugin for both pixi and lazer to make it available for everyone in html5gamedevs community. 

Look at the big guys - cocos2d-x. Or look at libgdx. We cant compete with them if we are separated, we are using tools that were created for their platforms (texturePacker, spine).

**Comment by mattstyles** [comment]
I'd assume that Pixi would remain a monorepo, but publish each component separately to npm (and/or wherever). Babel have recently restructured in such a way, and React is a monorepo, although there are many many examples. 

I'd have thought this approach could benefit a build system too.

**Comment by photonstorm** [comment]
@mattstyles yes exactly that.

To try and qualify my hastily written statement: the issue I have with StackGL is nothing to do with the fact it uses npm. It's purely the sheer volume of overhead that copying that structure for Pixi (or any sizeable library) would entail. StackGL is currently a massive 269 packages. It's fair to say that the vast majority of them have absolutely zero documentation of any kind (most have READMEs, next to none of them have jsdocs in the source, you literally have to guess what the parameters map to in lots of cases). The code structure and style between packages varies dramatically based on the author. Very few (if any?) of the packages tell you how they fit into the overall scheme of things, i.e. where in the flow they play a role (what they expect to receive as a valid object / type, and what they spit out again).

In short they expect you to have all of this knowledge up front, i.e. to know already perfectly well that what you really needed was a combine smooth shader to merge two distance fields, or to be capable of reading the source and implicitly understanding where to use them. This is 'hardcore GPU devs' land imho :) It's not for the masses. Heck it's beyond most advanced devs imho, those who actually want to ship something rather than scan though hundreds of packages hoping to find what they need.

So when I say you should avoid the StackGL structure approach I mean it purely from this point of view. The way it consists of lots of single-use functions and stand-alone components is actually perfect, it's just not 'consumer focused'. And the thought of having to maintain *hundreds* of npm packages, revisions, testing, dependency chains.. fuck me, that sounds like a full-time job right there, never mind actually coding anything!

@ivanpopelyshev don't misunderstand me, what you're doing for Pixi over the past few months is fantastic. Matt is a lucky guy to have you onboard :) When I say "it's not modular" I'm talking about the granularity of Pixi. It's currently very non-trivial if you wanted to totally replace say Pixi.Sprite (extend, yes, replace - no) because there are deep expectations throughout Pixi based on assuming that is what it will be asked to render.

A true 'renderer' would have no concept of a Sprite at all imho, you would just have to provide it a stack of vertices data and be done with it. It should almost be a pure texture renderer. But even then you could argue that the way it currently builds its interleaved vertices data structure based on GL_TRIANGLES is another 'assumption', yes? You yourself use GL_POINT for your tile map renderer I believe? But maybe my game would benefit from a tri strip instead? Or maybe I don't ever want to tint anything, so I'm adding in extra data I don't need to be there? (or maybe I want per vertices color blends, so I actually need to inject new data into the structure?!) - what Pixi does it does fantastically, but it does it in one very specific way, and trying to bend away from it is quite a task sometimes.

It's unrealistic of Pixi to cover all bases of course, but making some aspects of the actual renderer more independent could help massively. This isn't a simple case of chopping the files up a bit, it needs a new structural approach - and it's entirely possible it's not even worth doing, I mean there are only so many hours in the day, yes? If most people are already really happy with it, then why rock the boat? Change for changes sake is never a good reason.

I'm taking Lazer in this direction because I truly believe that is the way the web is heading. I guess it's very similar to the StackGL route, but monorepo'd and without its convention and docs drawbacks. *For me* "One fit's all" doesn't apply any more. If someone wants to dig into Lazer and utterly replace how I handle something as basic as a mousedown event, they should be able to do that. As long as I actually document my functions and explain how they link up, that is entirely possible too.

**Comment by mreinstein** [comment]
@photonstorm I guess I'm confused. It seems like your big beef with stackgl has more to do with lacking documentation and rigor. That seems to me like a separate problem, and bound to happen with any non-trivial sized project. If documentation isn't taken seriously and committed to no architecture is going to make that go away.

**Comment by ivanpopelyshev** [comment]
@photonstorm my tile renderer supports GL_TRIANGLES and it is actually just a renderer, you dump data there and clear it on scrolling, it doesnt care if you are using tiled or rpgmaker or whatevs. On top of that I can write library for tiled, rpgmaker or my own tile engine.

I think we need more structures like that in pixi, look closer to threejs. ParticleContainer was a good try in that direction. Its really pain to use ONLY standart pixi containers and sprites. 

Ok, now I understand your point.

But what about infrastructure? We can use same parsers/loaders and integration libraries in both projects even if they have different architecture. Are "texture" and "loader" classes really differ too much? I think if we move texture cache into separate module it will work for both projects. Things like [compressed textures]("https://github.com/pixijs/pixi-compressed-textures" )  can be separated into core module and adapters for pixi/lazer, built into two different js files.

**Comment by GoodBoyDigital** [comment]
Blimey! that this is a busy thread today!

Some really interesting points here to consider from all!  From a casual user point of view pixi feels pretty good (although a bit on the hefty side!) but a common running theme that I have encountered with more advanced users is this very issue of 'the pixi way or the high way'. Its great to see it being discussed here too.

Lets make sure v4 covers these users. Three.js is a great example in terms of how easy it is to use custom geometry and and shaders. By exposing ways to easily render things the way you want should be a core part of the v4 architecture (something we have already started playing around with!)

Theres a nice middle ground to be found between stack GL and pixi v3 that we should aim to hit with v4.

(along with all the other great suggestions!)

**Comment by grovesNL** [comment]
Most of the extensibility challenges can be solved by refactoring dependencies (Sprite, etc.) out of the renderer, which is a direction I think most would like to see. In fact the core of Pixi should just be the renderer, and pieces like the `InteractionManager` should be optionally layered on top. Personally I would like to be able to pass custom geometry and shaders just to the core, because it's something I am currently struggling with.

It is unfortunate that Phaser is moving away from Pixi because 2D WebGL is already somewhat of a tight niche. So many of the same efforts (resource, interaction, drawing primitives, animation, batching) become duplicated, instead of both teams working together to make a unified, extensible 2D framework.

**Comment by GoodBoyDigital** [comment]
@grovesNL totally agree :+1:

**Comment by mreinstein** [comment]
@grovesNL @GoodBoyDigital torwards reconciling commonalities and sharing more across teams, would it make sense to adopt a lower level API? Something like igloojs that would form less opinions about structure? https://github.com/skeeto/igloojs

Maybe there is a lower level common-ness that could be shared across teams. I agree that duplicating effort is painful.

**Comment by GoodBoyDigital** [comment]
I hoping that some of this stuff is already in place with some of the current webGL classes we use (such as the renderTarget / shader manager etc) and that it will be a case of tweaking / opening up what we have and adding stronger documentation to those elements. I will make sure to check igloo out too mind..

**Comment by englercj** [comment]
Whew, lots of things going on here! Thanks @photonstorm for that explanation, I not only understand, I agree. I'm really looking to make a serious shift in how pixi works with v4 to make it as usable by as many projects as we can.

When I say modularity I don't just mean being able to add/remvoe plugins I also mean our data model being modular, our objects plugable. If we make our own internal objects with the idea of them as plugins, then it should be easier for others to do the same to change something for their better.

**Comment by photonstorm** [comment]
@mreinstein 

> It seems like your big beef with stackgl has more to do with lacking documentation and rigor. That seems to me like a separate problem, and bound to happen with any non-trivial sized project. If documentation isn't taken seriously and committed to no architecture is going to make that problem go away.

It's definitely not the biggest issue (the fact it's an amorphous blob of packages that may or may not work together is imho), but it's a significant one. I bet if we were to break Pixi and Phaser down in the same way StackGL is (i.e. mostly a single function per file) then we'd get very close to the same number of packages they have. Yet we are fully capable of documenting our code and adhering to coding standards (mostly, nothing is 100%), so I really don't see any reason why StackGL shouldn't be capable of doing the same.

I can't help but think the reason it doesn't even try is maybe endemic of the npm mentality? Where it's a *rarity* to find well documented, maintained, consistent packages. I appreciate that npm as a delivery system has absolutely no bearing on this state of affairs at all - yet it's incredibly common. So maybe devs don't bother because they don't often see anyone else doing it?

@englercj 

> When I say modularity I don't just mean being able to add/remvoe plugins I also mean our data model being modular, our objects plugable.

Yes, exactly this. Having looked at igloo I'm not sure it's a great example actually, as it depends on its own structures internally. But what you said is absolutely right. A data driven design, not a class based one - that is the key right there.

There are challenges to this approach. One's I haven't yet figured out for myself. For example writing jsdocs for single-fire functions is actually extremely difficult when it comes to deciding where in the namespace they should exist. The last thing you want is a docs page which is just a list of a few thousand functions :)

Also for those who like using strongly typed environments you need to find a way to help them too. A TypeScript defs file for example usually covers a whole code base. But if you construct a code base where you know that maybe 50% of it is optional, do you include that in the defs? Maybe each module should have its own defs, but then that's a bit of a nightmare for TS users to deal with. I've yet to come up with a sensible solution for this, but I'm thinking maybe it involves using custom jsdoc tags and auto-generating TS defs when you build your package.

Lots of IDEs offer code-insight based on inspecting single js files for structure and jsdocs. So again they can only really do this based on a built package - which is a problem because maybe the part of the library the dev really wanted wasn't included in the default build?

A tangent - but also I've yet to see any project deal with internalisation of jsdocs.

So yes, lots of challenges! I feel they are all ones that have been solved elsewhere though, just need applying here.

As for the comments about duplicating effort between us all, that's a very good point. It does seem a bit pointless in some regards. I'm focused entirely on Lazer development for the next 3 to 4 months now, so I'll definitely keep these things in mind as I'm going ahead: about how best to share what I've dealt with, pass on learnings, and ultimately see if we can't provide code towards Pixi 4 and help with integration. I'd be more than happy to see Pixi4 in Lazer. It won't be the *only* renderer option, but it would still be great to have it as one.

**Comment by mreinstein** [comment]
@photonstorm ok yes I think we're on the same page. the problem is not that the code is modular, it's that these particular modules are not documented as well as they should be, lack tests, etc.

> So maybe devs don't bother because they don't often see anyone else doing it?

Part of the challenge of node and the npm ecosystem is finding modules (and authors) that tend to produce high-quality, documented work. For example, look at many of the packages from Mikola 

https://www.npmjs.com/~mikolalysenko

He generally puts  lot of effort into making his modules usable, with clearly defined interfaces. You can find garbage in any system, it's important to look at what can be done rather than the norm. :)

**Comment by photonstorm** [comment]
Of course, absolutely - you can find garbage (and shining examples) in nearly all walks of life. It's just that when you are building something that you know full-well cannot live in isolation, that is part of a bigger whole, and you don't even bother to document how it expects to play its part - well, in those cases you may as well not have bothered publishing it in the first place imho.

This is an interesting tangent but I know lots of devs look to three.js for inspiration because of its really high GitHub star count, etc - but from a docs point of view it's pretty terrible! Methods and properties are very rarely (if ever) documented at all, you literally have to understand what's going on 'as a whole' if you want to venture deeper than pulling apart the examples or tutorials. Thankfully the code is beautifully clean and makes this relatively easy, but it's still sorely lacking on the docs front!

The bigger and more complex the code base, the worse the docs tend to get. There is no denying the sheer power of libs like Goo Create, Turbulenz or BabylonJS - but the jsdocs for them all are utterly dire! Truly only the original developers can know how they work internally, knowledge surely kept in their minds because it's surely not committed to text :)

**Comment by andrevenancio** [comment]
I feel that this thread is more about structure than features :)

Please santa:
rounded corners for graphics.lineStyle <3

**Comment by englercj** [comment]
>  A data driven design, not a class based one - that is the key right there.

A whole bunch of this, I want this. All of it. I'm also thinking about data locality and speeding up access of sequences.

**Comment by activetheory** [comment]
Hey friends,

Thanks for all the hard work done on pixi -- it's an awesome tool that lets us do some creative stuff! I'd like to offer a few points of feedback based on what I've read in this thread.

1. Low level geometry/buffers/shaders
+10000. We've written some code that bolts custom shader functionality onto pixi, including shader cloning. This would be great to have as part of the framework as the code has changed a few times in subsequent releases and we have to re-digest the source and then update our code.

Additionally, being able to create a buffer for geometry paired with a custom shader would give maximum control and allow for some crazy results. The way three.js BufferGeometry works with ShaderMaterial is a good reference here.

2. As Matt mentioned, I'd strongly recommend perf testing Babel before committing 100%
ES6 is still not really supported by iOS and even Chrome is missing some important bits. We tested Babel and found that it creates some really, lets call them interesting, conversions when transpiling for ES5. I'm not super confident in its ES5 code on an application as performance critical as Pixi.

Thanks again!

**Comment by mreinstein** [comment]
I will probably get flamed for saying this but I still think coffeescript is the way to go. :) it's cleaner syntax, it supports classes, it has more features than are available in es6, and it generates es3 compatible output. :)

**Comment by drkibitz** [comment]
IMHO, before any of the pretty great ideas here...

:star2: perf benchmarking :star2: 

It will be a lot easier :palm_tree: , and we would pat ourselves on the back for how easy it would be to make any sort of decision on ideas in and out of this thread. We could have clear stats proving the way :steam_locomotive:.

**Comment by mattstyles** [comment]
Yeah, sorry @mreinstein, I think coffee would be a terrible choice, I generally hold this opinion for anything that deviates from spec, even if that spec has included a load of goodness from coffeescript (I'm on the fence for strongly typed languages like TS/Haxe as for large groups of devs they are very beneficial, still dont think it would be a good direction for Pixi though). Babel gives you the features you'd want and at some point you can drop it (this being the key bit), although I'd also share concerns over the performance of some of its transpilation (same for coffeescript). 

Of course part of the problem of transpilation is that browsers that *do* support those features dont get to use them, currently they are all slower than their ES5 counterparts anyway, but, that wont stay like that for long.

A limited subset of ES2015 features could be used, and I wonder if you could get to a situation where if you tighten browser support you wouldnt need to transpile at all, v4 is going to take a while after all. Unfortunately the features which are probably *most* beneficial to Pixi do not have good support.

@drkibitz perf testing is surely a must, totally agree that it should be heading up the list of priorities.

**Comment by mreinstein** [comment]
> I wonder if you could get to a situation where if you tighten browser support you 
> wouldnt need to transpile at all

Sure you could. In theory. In practice, transpilation never really goes away because the hot new sytnax features are always lusted after. After es6 becomes widespread, we'll want the stuff in es7. and es8. and so on. :p 

> v4 is going to take a while after all. 

I don't think we should take that as a necessity. I started this thread to gather ideas, but we shouldn't get carried away with implementing even 1/3 of it in v4. I think it's dangerous to fall into the the trap of having massive releases that take 6+ months. A more continuous release cycle is where it's at. :)

**Comment by mattstyles** [comment]
Yep, totally agree with you there. Keep the good stuff flying in to the project!

**Comment by photonstorm** [comment]
Perf tests for sure - although I'm always quite wary of them. There is a strong tendancy to test once, on a very limited range of device / browser combinations and then take that result as sacrosanct. Yet in reality Babel is releasing updates *daily*, so what may have held true last month almost certainly won't this month. Therefore making *architectural decisions* on perf testing something that volatile is always going to be fraught with assumptions (both ways).

**Comment by ivanpopelyshev** [comment]
@mreinstein No way we are building the core not in vanilla js. How correlates that and support of asm.js? However, coffeescript is good for projects like http://www.html5gamedevs.com/topic/18619-pixi-playground/ .

@mattstyles TS/Haxe/Dart, yay!

@drkibitz  And yes, we need that perf.

**Comment by mreinstein** [comment]
> No way we are building the core not in vanilla js. 

The fact that we're considering a transpiler like babel means we're already considering non-vanilla JS. hence the translation step. All I'm suggesting as devil's advocate is if we're going to consider transpilation, I'd be open to choosing a more aggressive source language. That's all. :)

> How correlates that and support of asm.js?

It doesn't, it's an incompatible idea. We've ruled that out when people pointed out asm.js style syntax doesn't mix well with non-asm.js code.

**Comment by ivanpopelyshev** [comment]
@mreinstein You have more experience than me with those things. Am I right that ES6 will become new vanilla soon? [offtop]yeah, and I'll write some headers for GWT[/offtop]

**Comment by mreinstein** [comment]
@ivanpopelyshev es6 will definitely become the new vanilla js. this chart gives a great overview of the current landscape  https://kangax.github.io/compat-table/es6/

You'll probably notice that there's an es7 tab too though. ;) The support there is not even remotely good. But how many of those features would we love to have (someday.)  I only bring this up as a point that transpilation is part of the mentality/philosophy of "I want to use the new shiny stuff today". So even if es6 becomes available everywhere, I suspect we'd still want to let pixi use advanced features that arent available yet. That's why I don't think it'll ever go away. Not unless the philosophy goes away with it.

Anyway I don't want to drag us into a quagmire of debate on philosophy. :) Just making a point to consider.

**Comment by mreinstein** [comment]
another architecture idea: replacing Point with a popular Vec2 module. This one is pretty good:

https://github.com/ianmcgregor/boid/blob/master/src/vec2.js

chainable, provides many operations, and has a built-in object pool. A lot of junk vec2 modules I've create a lot of spurious objects.

Another more radical suggesting: using arrays instead of objects to represent points. Again, don't want to start a flamewar but I know the entire stackgl crowd is all about the `[ x, y ]` syntax rather than `{ x: y:}`.

Again, devil's advocate..

 _ducks head out of way_

**Comment by ivanpopelyshev** [comment]
@mreinstein I agree. We can use pooling them in some lightweight elements (not in the core), particles and stuff. But for basic pixi containers pooling is bad. Is that vec2 used in box2d ports?

**Comment by grovesNL** [comment]
@edankwan's [pixi-custom](https://github.com/edankwan/pixi-custom) appears to have some basic steps towards create custom geometry and shaders that we may be able to learn from. It may be guidance for #2250 (@activetheory) also

**Comment by ivanpopelyshev** [comment]
Btw, does anyone know status of WebGL2?

**Comment by mreinstein** [comment]
@ivanpopelyshev in progress, and shipped nowhere :package: 

https://www.chromestatus.com/features/6694359164518400

**Comment by photonstorm** [comment]
@mreinstein I'd vote for arrays instead of objects (for Vecs, etc) - in fact it's perfectly possible to [have your cake and eat it](https://github.com/photonstorm/lazer/blob/master/src/math/vector/vec2/Vec2.js#L7). A note on that code: internally in Lazer I use arrays everywhere, but the Vec2 class is there for public consumption and structured in such a way that anyone using it can still exchange those objects with any Lazer function that is expecting a vec2 array and still work, with no conditional checks.

**Comment by mattstyles** [comment]
@mreinstein I'd vote for arrays too, but as a consumer of Pixi I probably shouldnt care too much for the implementation anyways. `sprite.setPosition` should just work as it currently does.

**Comment by mehmetminanc** [comment]
@mreinstein It seems like [it's quite straightforward](http://developer.humblebundle.com/post/112252930481/developing-for-asmjs-using-sdl2) to compile SDL2 as-is against asm.js. [In a subsequent post](http://developer.humblebundle.com/post/112074832801/evaluating-asmjs-for-your-game) they say that on asm.js graphics target is WebGL. I haven't tried it but the author sounds like they figured it out.

My $.02 on architectural discussion, not that I think you should care, just checked StackGL, scared the hell out of me. I believe more granular modules could be useful and helpful, but module-per-function is definitely not dev-friendly.

**Comment by GoodBoyDigital** [comment]
Absolutely agree! 

I also agree with the comments of not having to do a complete teardown. One awesome new feature at a time :)

After a lot of thinking and digging around, architecturally I believe we can reach our goal by reshuffling existing code within pixi. I am currently investigating 'step one' which is to separate out the rendering code from the model code.  I am also confident that this can be achieved without many (or any!) Changes to the top level API. Results so far are encouraging! Watch this space for updates...

**Comment by ivanpopelyshev** [comment]
@GoodBoyDigital that's the right way. I can help to review your changes as soon as you make some. Which branch will it be?

**Comment by GoodBoyDigital** [comment]
Awesome! Thanks @ivanpopelyshev! Definitely keen to get brains on it :+1: 
I will push up the branch once its ready. the branch will be 'dev-refactor'

**Comment by grovesNL** [comment]
@GoodBoyDigital Sounds great, looking forward to it!

**Comment by englercj** [comment]
> After a lot of thinking and digging around, architecturally I believe we can reach our goal by reshuffling existing code within pixi.

I'm not sure I feel the same way, I don't think we can reach a data-driven rendering system by just moving around existing code. I strongly believe this is a large shift in the way pixi works, its going to require some serious rethinking and rewriting.

**Comment by GoodBoyDigital** [comment]
@englercj Totally understand, maybe i was a bit fast and loose with the term 'reshuffling existing code'. There will certainly be chunks of rewriting and rethinking. We will work together to ensure we are on the same page too! Once I have this first pass ready I am sure it will explain my thinking in a much stronger way.

**Comment by ivanpopelyshev** [comment]
I think we need combination of "entity-component" and mixins. Heavy configurable containers and lightweight sprites.

Current pixi approach is "Container is the man". We can put gigantic number of sprites (its our "data") into different type of containers and receive different results in performance and graphics quality.

Pixi default container is the beautiful mixin of
1) transform calculator (see updateTransform)
2) interaction stuff
3) caching stuff (cacheAsBitmap)
4) filters stuff

And I think default behaviour needs to be that way, but we need de-couple that, so other containers will contain only some of them or some additional stuff. For example, ParticleContainer doesnt contain filters code, because it overrides "renderWebGL". 

Pixi default Sprite:
1) all container stuff
2) sprite logic
3) interaction (containsPoint)
4) renderCanvas
5) updateTransform

Its too heavy. Of course we need that version, because that API is universal and awesome. Imagine that we'll add polygon packing, Sprite will become even more heavy! 
We need lightweight sprite to be used for particles, which contains only arguments for shader.

**Comment by mreinstein** [comment]
bringing the power of SIMD to gl matrix

https://hacks.mozilla.org/2015/12/bringing-the-power-of-simd-js-to-gl-matrix

Firefox is the only browser that supports this but they seem to have gotten some good optimizations out  of this in a few places.

**Comment by ivanpopelyshev** [comment]
@mreinstein yeah, this thing we need too. Supported by only firefox nightly but I'm ready to include this both into pixi and pixi-spine.

**Comment by mreinstein** [comment]
> I believe more granular modules could be useful and helpful, but module-per-function is definitely not dev-friendly.

module-per function or several modules per-function is not super relevant. The biggest factors in how usable a module is:
* how well documented
* how consistent the APIs are between related modules
* how well defined the test suite is

The way I've best heard it described by is following the UNIX mantra "do one thing and do it well". Your module should do one specific thing and do it as cleanly and simply as possible. It shouldn't be a grab-bag of functionality. @substack has a great blog entry on this, highly recommended http://substack.net/finding_modules

**Comment by ivanpopelyshev** [comment]
One more time, I just want to say that **pixi default behaviour is awesome** and I dont want to change it, and people love pixi because of it. But de-coupling needs to be done so we'll be able to use more behaviours. I have so many lines of code in bomber dedicated to optimizations and tricks that can work only in certain conditions, and all that stuff can be made into special kind of containers of components in those containers.

**Comment by GoodBoyDigital** [comment]
> I just want to say that pixi default behaviour is awesome and I dont want to change it, and people love pixi because of it.

On the money my friend! Pixi's popularity is due to a nice simple to use front end. I am keen to preserve this. Its fair game behind the scenes though. 

Lets keep the ideas flying!

**Comment by mreinstein** [comment]
All of the features/ideas we've been proposing will grow the size of pixi. Given how busy @GoodBoyDigital  @englercj @drkibitz and @ivanpopelyshev are already, I'm wondering if there are things we could do to *reduce* the scope of pixi in the long run.

A (possibly) controversial proposal: Drop canvas support altogether in future versions, but support it for a while in pixi v3.  Bullet points around this idea:

* when pixi first made it's debut, webgl was not at the state of availability that it currently is. http://caniuse.com/#feat=webgl
* maintaining a swappable rendering backend is not a small feat. I get that this is already working quite well, but if we were to focus future pixi to just be a super tight webgl based 2d rendering engine if it would reduce the maintenance and new feature costs significantly over time. A lot of unnecessary abstractions go away.
* besides having better support across platforms (for now) and a handful of blend modes, does canvas offer anything else over webgl?
* the canvas API is simple enough that I've never seen a massive benefit from going with pixi over just writing straight-up canvas calls.

I get that people are using pixi now on platforms that don't have stellar webgl support yet, but I also think this is a temporary situation. We could do something like jquery did when they stopped supporting ie < 9. jquery 1.x maintained further backwards compatibility, and 2.x stripped support. maybe pixi v4 is a realization that the web will soon have ubiquitous webgl and we shed canvas, but keep maintaining it in v3, just without fancy new features.

**Comment by ivanpopelyshev** [comment]
@mreinstein leave it as a plugin, but with full backwards support in it. I can try to gather stats of how many bomber users have webgl enabled.

**Comment by GoodBoyDigital** [comment]
I agree it will still need to be there. But as @englercj suggested at the top also, canvas no longer needs to be a first class citizen.

**Comment by ivanpopelyshev** [comment]
Dont forget that some major things can be achieved only with canvas - very thin lines as an example.

**Comment by ivanpopelyshev** [comment]
@GoodBoyDigital can you point to that open-source project with Adobe CC?

**Comment by ivanpopelyshev** [comment]
I just discovered this thing: http://dragonbones.github.io/ . And this too: http://www.egret.com/ .

**Comment by ivanpopelyshev** [comment]
MathBox2 is powered by https://github.com/unconed/shadergraph

**Comment by kittykatattack** [comment]
> Pixi's popularity is due to a nice simple to use front end

Thank you, as a Pixi end-user this is very re-assuring to hear!

**Comment by FlorianLudwig** [comment]
@mreinstein 

> the canvas API is simple enough that I've never seen a massive benefit from going with pixi over just writing straight-up canvas calls.

Yes it does: Good antialiasing across browsers. We have been using PIXI.Graphics to draw on a hidden canvas and use it as texture instead of using PIXI.Graphics on the WebGL renderer as the quality ends up much better in some browsers.

Also we have seen more crashes and render bugs with the WebGL backend so dropping the canvas renderer as fallback would mean we need to stick to v3.

**Comment by englercj** [comment]
> Also we have seen more crashes and render bugs with the WebGL backend so dropping the canvas renderer as fallback would mean we need to stick to v3.

This assumes that the WebGL renderer in v4 would suffer from the same issues as v3. Unless you are referring to crashes not related to pixi's implementation?

**Comment by ivanpopelyshev** [comment]
There are webgl crashes not related to pixi implementation. Canvas should be separated from core, but it must be available in standart pack. Even if pixi doesnt support "dirty rect" optimization which is critical for previous gen of html5 engines, we must support canvas.

**Comment by FlorianLudwig** [comment]
@englercj It is getting offtopic but I never really dug deep into it but AFAIK the crashes where not pixi-related since some go away by reloading the page. In any case - whatever you do in js it should not crash your browser ;) If it does it is a browser bug, imho.

@ivanpopelyshev agree.

We also use pixi in node.js using node-canvas which is an important usecase for us - so another dependency on the canvas renderer for us.

**Comment by ivanpopelyshev** [comment]
@FlorianLudwig high five me brother! node-canvas and pixi, server rendering rules :+1:

**Comment by mreinstein** [comment]
earlier someone was asking for some benchmarks on the fancy  expirimental feautures coming out. The unity team put together a pretty solid look at usage of asm.js, shared array buffers, edge, and firefox 64 bit.  http://blogs.unity3d.com/2015/12/15/updated-webgl-benchmark-results/

impressive...

**Comment by mattstyles** [comment]
I find those unity tests a little strange (if not just misleading), certainly my own experience using webGL Pixi sees regular Firefox running at roughly 50-75% the frames of Chrome (Mac - nVidia). The unity tests, even on regular ole Firefox are coming out much better than Chrome. Most of my testing has also included P2 so it could be the amount of JS execution that is the problem for FF, but, still, a little strange. This is backed up by comments on the forum where other users have similarly found FF webGL to be flaky.

**Comment by andrevenancio** [comment]
Another feature that could be quite nice is pixi native pixel collision detection!

**Comment by Jorisslagter** [comment]
I am using Pixi.js for my work daily, and I am following this topic since the beginning (two weeks now?). I am not sure if I am supposed to mix in and request some features my self. A lot of thinks I use or needed, and now requesting, I found effectually somewhere on the internet. But I thought it was a little bit sad, that it wasn't in Pixi.js natively. 

## I hope Pixi.js V4 will support

* z-indexes
* easing
* timers (countdowns, intervals and timeouts)
* UI behavior (clamping, spacing elements evenly, input's, etc)

## Sharing
And as for the community. I would love to see some exchange environment and resources sharing platform dedicated to Pixi.js. 

## Learning
I am sure some of you know Laravel and it's dedicated netflix-a-like video's? It's training video's are for professionals. They are at high quality, about the architecture, design principles and about code in general. I trying not to promote it, but purely illustratie my wish to see some kind of platform, dedicated for Pixi.js professionals. 

I also would love to learn more about game development with Pixi.js. As this far, it's somewhat lonely and passive learning curve. I am trying to mix in to the community, but whois taking my hand?

**Comment by whizzkid** [comment]
just ringing in with my 2 cents,

We make e-learning games for highschools in the Netherlands, where our games are used inside schools. Unfortunately those schools are notoriously bad at upgrading, so we still HAVE TO support old browsers that do not support WebGL. (and we can't force them to upgrade, we've tried... )
The games are also played over a range of mobile devices, lots of which simply do not support webGL.

So for us, canvas rendering is hugely important.

**Comment by photonstorm** [comment]
@whizzkid I'd say that for most people using Pixi _professionally_ canvas rendering is more important than WebGL. I'd be very surprised if this was ever 'dropped', although I do agree completely with it being made optional (in the same way that WebGL should be optional as well)

**Comment by andrevenancio** [comment]
@photonstorm which data are you relying on to say that canvas rendering is more important than WebGL? This is shocking :)

**Comment by photonstorm** [comment]
@andrevenancio every client I've ever had in the past 4 years of making HTML5 games. And those I'm working with at the moment. The biggest ones out there (Disney, BBC, Marvel, etc) need the fallback, as for a sad percentage of their traffic it's actually the only version any of them get to see.

**Comment by drkibitz** [comment]
Some v4 goals:
- choose a development methodology, and stick with it
- when new code is written, make sure it is written with tests or testing in mind
- when old code is fixed, write tests for it
- did I mention write tests?
- TDD

This is JavaScript, and the very nature of the language almost requires tests to cover code that is meant for a high-level of consumption. Especially code that has a lot of platform abstractions, variance, and unknowns; as well as a lot of internal variance with plugins. I have some be very blunt here and I say, I have no idea how PIXI has made it this far with the amount of tests it doesn't have, and the importance the core maintainers and community do not put on fixing this situation. All that it needs is a plan of attack, and everyone who is all excited about new features, refactoring, and what-not should be involved. Because if you want a new feature, it should be required that tests come with it. The same goes for refactoring. No, I don't like writing tests either. I've only met about 5 people in my career who do, but a lot more just bite the bullet and do it anyway, because it makes maintaining code-bases soooooo much easier in all aspects of being an open source software maintainer.

**Comment by drkibitz** [comment]
BTW I'm guilty too, but that might be because I tried to get a suite started, then the refactor happened, then it happened again. Now here we are talking about it again.

If refactoring always wins over testing, and testing is never a part of refactoring...

Where does that leave you?

**Comment by GoodBoyDigital** [comment]
Totally agree, for me it has always come down to time. The little time I have is spent on pushing pixi forward but I do agree that testing is important. I think what I really need is a bit of an education in how to write them too (eg how to test webGL output). @drkibitz once the refactor is complete, would you be willing to help us out in setting up some tests?

**Comment by mreinstein** [comment]
One of the most important tenets of TDD is that you write the tests first, and the code to make those tests pass second. The idea behind this is you come up with the most valuable test cases around the same time that you're conceptualizing the functionality to be built. If we're bolting tests onto an existing codebase, that's not really TDD.

TDD is great, but it takes a **metric fucktonne** of resources to implement. Perhaps it could be implemented for some core critical modules but I just don't see it being realistic to implement across the whole codebase for a dev team that is already seriously over constrained and resource starved.

I dunno, maybe it's because I haven't been bitten by many bugs in pixi. It mostly just works as advertised. I'm hesitant to suggest that the core maintainers should switch from building sorely needed features (webgl2, real lighting primitives, etc) and switch over to massively expanding the test suite.

**Comment by drkibitz** [comment]
@mreinstein It should come down to what PIXI is to the majority of the community, in comparison the the already abundant PIXI-like tools out there. I'm asking to try to look at things through that particular lense, because if PIXI is a constant moving target of refactored features that continues to be untested, there either already is, or there is definitely going to be a better option than it. If the acronym TDD frightens, then I will renounce that point, but still replace it a point of simply asking for and open mind, to the idea that tests are not a waste of time, not during before, during, or after a refactor, and it doesn't matter who is doing them because it is everyone's responsibility. Talking about resources, there's over 9k stars, 630 watchers, and a community of developers all swirling around PIXI at various levels of interest. If the core developers do not take testing seriously, nether will they.

Lets look at it how things are, and have been done, as is if they will continue with the v4.
One person writes support for webgl2, testing it either locally, with an example, or with an example derived from a proprietary codebase that is for a client. The same thing is done for lighting primitives but by a different person, who tests things locally in a different way, most likely with a different semi-private example. This will be the state of things as they are merged into the open source version after those 2 developers, with the exception of maybe a couple more, talking amongst themselves. After the mege, another person or two (out of the 630+ watcher mind you), try it out, file issues, make pull requests, and comment on commits directly. This process happens for quite a while, until 1-4 people think the library is relatively bug free based on a perusal of issues, and how stable they believe things are with a general look at the examples, the docs, personal use cases, and how quite things are on github. After some time that version is released by Chad, the rest of the community uses it for a while, and after some more time, people wonder why things were done in a certain way, and a thread is started to talk about possible new directions, but ultimately what comes out of that thread is a repeat of what was just described.

**Comment by mreinstein** [comment]
> If the acronym TDD frightens, then I will renounce that point

@drkibitz I've managed a 15 person development team and personally introduced TDD as the primary development style. You're preaching to the choir! :) You don't need to sell me on it's merits, I'm already onboard with the strategy; it works well for many types of products.

>  asking for an open mind, to the idea that tests are not a waste of time, not before, during, or after a refactor, and it doesn't matter who is doing them because it is everyone's responsibility.

No one is disputing that tests are great.  When code is written first without a notion of unit testing it, you'll find that significant refactoring has to happen just to make the code fit into an interface that is testable. It's not insignificant work.

Anyway, I don't want this to spiral into a political debate. We should stick to the list of what we *could* do, and let the core maintainers decide what they want to work on.

In fact, @GoodBoyDigital  it might be a good time to lock the thread. New ideas stopped flowing in 3 weeks ago.

**Comment by ramsundhar20** [comment]
I think, We must be moving to Typescript. Similar to Angular 2. Unlike earlier, JavaScript is evolving faster, so does the optimizations of the language. By using a pre compiler like typescript, we can seamlessly be in interoperability with latest advancements.

**Comment by GordoRank** [comment]
The feature I would love to see most of all is the ability to use the new ``OffscreenCanvas``` for both WebGL/Canvas.  This enables all drawing/rendering work and JS computation to be completed in a webworker.  Leaving the main browser UI thread free.

https://developer.mozilla.org/en-US/docs/Web/API/OffscreenCanvas

You basically transfer the context ownership to a webworker.  Do all the rendering/JS work in the worker.  Then finally, commit it to the context to draw it.  This even means you can theoretically run an extremely heavy duty image processing WebGl shader that takes 20 seconds to complete, from within a webworker while keeping the main thread free.

This is my opinion is the biggest thing to hit web graphics for some time.  Even if actual support can't be ready for v4, I believe it should be considered and kept in mind, while designing pixi's structural changes etc.

Currently all major browser vendors have agreed on the spec and intent to implement.

Firefox has a completed implementation behind an experimental flag, and chrome development is well underway.  As for Edge, i'm not sure where they are with it yet.

It should be extremely easy to detect ```OffscreenCanvas``` support and run PIXI's rendering pipeline through a webworker as long v4 considers it structurally.

Anyway, just some thoughts.

**Comment by GordoRank** [comment]
Here is some further reading form Mozilla regarding the WebGL in webworkers.

https://hacks.mozilla.org/2016/01/webgl-off-the-main-thread/

**Comment by GoodBoyDigital** [comment]
thanks for all the info and ideas everyone. We have enough to get cracking with v4 now :+1:
